### PR TITLE
Add a configuration option to disable enforcements of constraints.

### PR DESCRIPTION
**Pull request by hlinnaka**
Add a new EnforceConstraints hint configuration option. If it's set,
ORCA will not add Assert nodes to enforce CHECK and NOT NULL constraints
on INSERT and UPDATE statements.

This is useful for GPDB, which is prepared to enforce the constraints on
its own. In theory, the optimizer could prove the inserted row to always
satisfy the constraints, in which case it could optimize away the checks,
whereas the executor can't do that on its own. But ORCA doesn't currently
attempt to do that. The reason I'd like to enforce the constraints in
GPDB side instead of as Assertion nodes, is that you get a different error
message from the assertion node. That's annoying, because we have to
maintain alternative expected output files for tests that hit CHECK or
NOT NULL constraints because of that.

Bump version number, since this is an incompatible change.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Any objections to the general idea?

Two questions on the implementation:

1. Is CHint the right place for this flag? I couldn't find any place for configuration options like this, that are not merely hints or tuning parameters for what kind of plans ORCA will create. This is not really a hint..

2. How should this be tested? Once we make the GPDB changes to set the flag, then the GPDB regression suite will cover this. But do you want something in ORCA's own test suite for this?

**Comment by vraghavan78** [comment]
-1. Why not:

* Change the error message to match it to the planner
Or
* Remove the assert during CTranslatorDXLToPlStmt for insert/update?

I am against this change, if the host system wants something it should be done at the host system level and not in GPORCA

**Comment by vraghavan78** [comment]
If you decide option 2 the fix should go in CTranslatorDXLToPlStmt::PplanDML

**Comment by hlinnaka** [comment]
On 10/16/2017 11:35 PM, Venkatesh Raghavan wrote:
> -1. Why not:
> 
> * Change the error message to match it to the planner

The error message is constructed deep down in ORCA. Seems difficult to 
always keep it in sync with the current GPDB version.

In later versions of PostgreSQL the error message will also include 
details of the failing row, like this:

postgres=# create table checktest (positive int4 check (positive > 0), x 
text);
CREATE TABLE
postgres=# insert into checktest values (-1, 'foo');
ERROR:  new row for relation "checktest" violates check constraint 
"checktest_positive_check"
DETAIL:  Failing row contains (-1, foo).

> Or
> * Remove the assert during CTranslatorDXLToPlStmt for insert/update?

Hmm. That might work, if we can assume that the Assertion nodes are 
always immediately below the DML node. And that there are no other 
assertion nodes than those generated for CHECK and NOT NULL constraints, 
i.e. that we can safely remove any assertion nodes that we see. Are 
those safe assumptions? Makes me a bit uneasy to modify the plan that 
ORCA generated like that, though.

> I am against this change, if the host system wants something it should be done at the host system level and not in GPORCA
I don't understand what you mean by that principle. It feels only 
natural to me that the host system can tell ORCA what capabilities and 
characteristics the executor has. Including whether the executor will 
enforce constraints on its own, or if it needs Assertion nodes in the plan.

- Heikki

**Comment by bhuvnesh2703** [comment]
Please help us understand the reason for below
- Why we want to leave the option to set the enable constraint flag to false? If we run GPORCA alone, it should produce the Assert node?
- If we plan to set the flag in GPDB so that we don't generate the Assert node, we can make the changes in translator DXL to PlStmt? Then we will not need to add any hint configuration and it will be more GPDB specific.


Also, Does it make sense to only change the error message instead of adding these checks?
Current error message:
```
pivotal=# insert into foo1 values (1,1, NULL);
ERROR:  One or more assertions failed  (seg0 127.0.0.1:25432 pid=50848)
DETAIL:  Not null constraint for column c of table foo1 was violated
pivotal=# set optimizer=off;
SET
Time: 0.723 ms
pivotal=# insert into foo1 values (1,1, NULL);
ERROR:  null value in column "c" violates not-null constraint  (seg0 127.0.0.1:25432 pid=50848)
```

Here is a patch for error message change and the results.
```
In ORCA
diff --git a/libgpos/src/error/CMessage.cpp b/libgpos/src/error/CMessage.cpp
index b1c902617f..470e914702 100644
--- a/libgpos/src/error/CMessage.cpp
+++ b/libgpos/src/error/CMessage.cpp
@@ -252,7 +252,7 @@ CMessage::Pmsg

                CMessage(CException(CException::ExmaSQL, CException::ExmiSQLNotNullViolation),
                                 CException::ExsevError,
-                                GPOS_WSZ_WSZLEN("Not null constraint for column %ls of table %ls was violated"),
+                                GPOS_WSZ_WSZLEN("null value in column \"%ls\" violates not-null constraint"),
                                 2, // column name, table name
                                 GPOS_WSZ_WSZLEN("Not null constraint for table was violated")),


In GPDB
diff --git a/src/backend/executor/execMain.c b/src/backend/executor/execMain.c
index 243c19bc3f..9397c0841b 100644
--- a/src/backend/executor/execMain.c
+++ b/src/backend/executor/execMain.c
@@ -4082,7 +4082,8 @@ ExecConstraints(ResultRelInfo *resultRelInfo,
                                slot_attisnull(slot, attrChk))
                                ereport(ERROR,
                                                (errcode(ERRCODE_NOT_NULL_VIOLATION),
-                                                errmsg("null value in column \"%s\" violates not-null constraint",
+                                                errmsg("One or more assertions failed"),
+                                                errdetail("null value in column \"%s\" violates not-null constraint",
                                                NameStr(rel->rd_att->attrs[attrChk - 1]->attname))));
                }
        }

```

Results after error message change:
```
pivotal=# set optimizer=on;
SET
Time: 0.765 ms
pivotal=# insert into foo1 values (1,1, NULL);
ERROR:  One or more assertions failed  (seg0 127.0.0.1:25432 pid=57755)
DETAIL:  null value in column "c" violates not-null constraint
pivotal=# set optimizer=off;
SET
Time: 0.720 ms
pivotal=# insert into foo1 values (1,1, NULL);
ERROR:  One or more assertions failed  (seg0 127.0.0.1:25432 pid=57755)
DETAIL:  null value in column "c" violates not-null constraint
pivotal=#
```

Regarding adding tests:
- If the change is in translator or in GPDB, i think the tests in gpdb icg are a good place for that.
- If there is no change required in ORCA, then we may skip adding any tests in ORCA as we already have mdp files with assert related checks.

**Comment by vraghavan78** [comment]
>
> I don't understand what you mean by that principle. It feels only
> natural to me that the host system can tell ORCA what capabilities and
> characteristics the executor has. Including whether the executor will
> enforce constraints on its own, or if it needs Assertion nodes in the plan.
>

You make a valid point. Let me think again.
ᐧ

**Comment by vraghavan78** [comment]
@bhuvnesh2703 @dhanashreek89 what changes in the error messages need to happen for CHECK constraints if we went that approach?

**Comment by hlinnaka** [comment]
On 10/17/2017 12:14 AM, Bhuvnesh wrote:
> Please help us understand the reason for below
> - Why we want to leave the option to set the enable constraint flag
> to false? If we run GPORCA alone, it should produce the Assert node?
You mean, we could just change ORCA to never produce Assert nodes? I 
assumed that we could not do that, to support HAWQ or older versions of 
GPDB. If we can, that's great!

> - If we plan to set the flag in GPDB so that we don't generate the
> Assert node, we can make the changes in translator DXL to PlStmt?
> Then we will not need to add any hint configuration and it will be
> more GPDB specific.
You mean, just strip out any Assert nodes in the translator? Works for 
me. I was afraid that ORCA might produce Assert nodes for other 
purposes, where removing them would not be cool.

> Also, Does it make sense to only change the error message instead of adding these checks?

I don't want to change the error message in GPDB. I want to keep the 
error message identical to the upstream PostgreSQL one, to avoid silly 
merge conflicts.

We could change the error message in ORCA, but then we'd need to keep it 
in sync with the current GPDB/Postgres version, which seems tedious.

- Heikki

**Comment by vraghavan78** [comment]
@bhuvnesh2703 we need to keep upstream conflicts to the bare minimum. So if we can change Orca error messages to match Planners that would be ideal.

**Comment by bhuvnesh2703** [comment]
@hlinnaka 
- In Older GPDB version or HAWQ, we would be using a specific version of ORCA. So if we make the change in ORCA now, i believe none of them should be impacted. One for sure, GPDB older version will not use the latest ORCA. HAWQ is on very old ORCA, if they plan to use current ORCA it will be a minor change to incorporate the changes being done in GPDB to support this constraint enforcement alongside the bigger refactoring to bump to use latest orca.
- What all will be impacted if we change that in translator: Need to investigate that and find out what all cases we put the assert node.
- Changing the error messages in ORCA seems fine. I hope they would not be changed enough in upstream that it will become a regular task. Though i do agree that, it might require changes when changed in upstream. Also it would not be that pressing as this will impact only the ORCA answer files and we can change it when needed.

Gave a quick try, may be we can go this path of changing the error message:
`CheckForAssertViolations` seems to be GPDB specific as its not available in Postgres, so we can make the desired changes here i guess. (Not sure if there is any other assertion which does not have `errorString.data`, otherwise the message `One or more assertions failed` is useless)
```
In GPORCA
diff --git a/libgpos/src/error/CMessage.cpp b/libgpos/src/error/CMessage.cpp
index b1c902617f..470e914702 100644
--- a/libgpos/src/error/CMessage.cpp
+++ b/libgpos/src/error/CMessage.cpp
@@ -252,7 +252,7 @@ CMessage::Pmsg

                CMessage(CException(CException::ExmaSQL, CException::ExmiSQLNotNullViolation),
                                 CException::ExsevError,
-                                GPOS_WSZ_WSZLEN("Not null constraint for column %ls of table %ls was violated"),
+                                GPOS_WSZ_WSZLEN("null value in column \"%ls\" violates not-null constraint"),
                                 2, // column name, table name
                                 GPOS_WSZ_WSZLEN("Not null constraint for table was violated")),

In GPDB
diff --git a/src/backend/executor/nodeAssertOp.c b/src/backend/executor/nodeAssertOp.c
index ab66223302..e14e5335d5 100644
--- a/src/backend/executor/nodeAssertOp.c
+++ b/src/backend/executor/nodeAssertOp.c
@@ -83,8 +83,9 @@ CheckForAssertViolations(AssertOpState* node, TupleTableSlot* slot)
        {
                ereport(ERROR,
                                (errcode(plannode->errcode),
-                                errmsg("One or more assertions failed"),
-                                errdetail("%s", errorString.data)));
+                                errmsg("%s", errorString.data)
+                                ));
```

@vraghavan78 
For check constraint we could do a similar change and change the message produced by orca.
```
pivotal=#  insert into foo2 values (1,1,1);
ERROR:  Check constraint foo2_c_check for table foo2 was violated  (seg0 127.0.0.1:25432 pid=66121)
pivotal=# set optimizer=off;
SET
Time: 0.918 ms
pivotal=#  insert into foo2 values (1,1,1);
ERROR:  new row for relation "foo2" violates check constraint "foo2_c_check"  (seg0 127.0.0.1:25432 pid=66121)

We can change it here:
		CMessage(CException(CException::ExmaSQL, CException::ExmiSQLCheckConstraintViolation),
				 CException::ExsevError,
				 GPOS_WSZ_WSZLEN("Check constraint %ls for table %ls was violated"),
				 2, // constraint name, table name
				 GPOS_WSZ_WSZLEN("Check constraint for table was violated")),
```

BC & EK

**Comment by hlinnaka** [comment]
On 10/17/2017 01:49 AM, Bhuvnesh wrote:
> - In Older GPDB version or HAWQ, we would be using a specific version
> of ORCA. So if we make the change in ORCA now, i believe none of them
> should be impacted. One for sure, GPDB older version will not use the
> latest ORCA. HAWQ is on very old ORCA, if they plan to use current
> ORCA it will be a minor change to incorporate the changes being done
> in GPDB to support this constraint enforcement alongside the bigger
> refactoring to bump to use latest orca.
Ok. In that case, I propose that we rip out the Assertion node stuff 
altogether, and make it always the executor's responsibility enforce 
constraints. That's a nice simplification. Any objections?

- Heikki

**Comment by bhuvnesh2703** [comment]
@hlinnaka i would suggest that changing the error message seems to be the minimal change without any impact or downside of it. So, i would incline towards not ripping it out and making the changes in executor. Do you see any pressing impacts of the error message change, as the patch suggested above does not conflict with upstream and does not change any functionality either?

**Comment by hlinnaka** [comment]
On 10/17/2017 08:45 PM, Bhuvnesh wrote:
> @hlinnaka i would suggest that changing the error message seems to be
> the minimal change without any impact or downside of it. So, i would
> incline towards not ripping assert node code and making the changes
> in executor. Do you see any pressing impacts of the error message
> change, as the patch suggested above does not conflict with upstream
> and does not change any functionality either?

I would prefer to rip it out, because then we have less code to 
maintain, and we will not need to worry about keeping the error message 
in sync anymore.

It's not pressing, but I would rather bite the bullet now, than take on 
the technical debt.

Turning it around, any reason *not* to rip it out now?

- Heikki

**Comment by vraghavan78** [comment]
After some more thought, I am fine with GPDB passing this as a hint. I am against ripping the whole thing because that make it too tightly couples with GPDB/Postgres. We need to add tests. We have take one of  the DML tests and pass the hint to avoid constraints for DML. You can add it to DML tests. One for Insert and one for update would statement would suffice.

File: libgpopt/include/gpopt/engine/CHint.h [hunk 1]
```diff
@@ -107,6 +111,14 @@ namespace gpopt
 				return m_ulBroadcastThreshold;
 			}
 
+			// If true, ORCA will add Assertion nodes to the plan to enforce CHECK
+			// and NOT NULL constraints on inserted/updated values. (Otherwise it
+			// is up to the executor to enforce them.)
+			BOOL FEnforceConstraints() const
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/engine/CHint.h]: FEnforceConstraintsOnDML to be explicit.

File: libgpopt/include/gpopt/engine/CHint.h [hunk 2]
```diff
@@ -118,7 +130,8 @@ namespace gpopt
 										INT_MAX, /* ulJoinArityForAssociativityCommutativity */
 										INT_MAX, /* ulArrayExpansionThreshold */
 										JOIN_ORDER_DP_THRESHOLD, /*ulJoinOrderDPLimit*/
-										BROADCAST_THRESHOLD /*ulBroadcastThreshold*/
+										BROADCAST_THRESHOLD, /*ulBroadcastThreshold*/
+										true /* fEnforceConstraints */
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/engine/CHint.h]: fEnforceConstraintsOnDML

File: libnaucrates/include/naucrates/dxl/xml/dxltokens.h [hunk 3]
```diff
@@ -61,6 +61,7 @@ namespace gpdxl
 		EdxltokenArrayExpansionThreshold,
 		EdxltokenJoinOrderDPThreshold,
 		EdxltokenBroadcastThreshold,
+		EdxltokenEnforceConstraints,
```
Comment (vraghavan78) [inline libnaucrates/include/naucrates/dxl/xml/dxltokens.h]: ...EnforceConstraintsOnDML

File: libnaucrates/src/xml/dxltokens.cpp [hunk 4]
```diff
@@ -84,6 +84,7 @@ CDXLTokens::Init
 			{EdxltokenArrayExpansionThreshold, GPOS_WSZ_LIT("ArrayExpansionThreshold")},
 			{EdxltokenJoinOrderDPThreshold, GPOS_WSZ_LIT("JoinOrderDynamicProgThreshold")},
 			{EdxltokenBroadcastThreshold, GPOS_WSZ_LIT("BroadcastThreshold")},
+			{EdxltokenEnforceConstraints, GPOS_WSZ_LIT("EnforceConstraints")},
```
Comment (vraghavan78) [inline libnaucrates/src/xml/dxltokens.cpp]: again add OnDML

File: libgpopt/include/gpopt/engine/CHint.h [hunk 5]
```diff
@@ -45,6 +45,8 @@ namespace gpopt
 
 			ULONG m_ulBroadcastThreshold;
 
+			BOOL m_fEnforceConstraints;
+
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/engine/CHint.h]: add OnDML to the end to make it clear that this is only just below DML not in other scenarios that Orca adds an assert.

**Comment by vraghavan78** [comment]
Use https://github.com/greenplum-db/gporca/blob/master/data/dxl/minidump/UpdateNotNullCols.mdp as an example

**Comment by vraghavan78** [comment]
Example for hints is: https://github.com/greenplum-db/gporca/blob/a4e68b5fd0b42b07fe52e44d539595528eaec15f/data/dxl/minidump/ConstTblGetUnderSubqUnderProjectNoOuterRef.mdp

**Comment by vraghavan78** [comment]
example for Insert is https://github.com/greenplum-db/gporca/blob/a4e68b5fd0b42b07fe52e44d539595528eaec15f/data/dxl/minidump/InsertNotNullCols.mdp

**Comment by hlinnaka** [comment]
On 10/18/2017 07:19 AM, Venkatesh Raghavan wrote:
> After some more thought, I am fine with GPDB passing this as a hint.
> I am against ripping the whole thing because that make it too tightly
> couples with GPDB/Postgres.

ORCA is very tightly coupled with GPDB/Postgres already. Ripping out the 
Assert node stuff wouldn't move the needle in any meaningful way. I 
don't like the idea of leaving behind dead code. Even if we add tests, 
sooner or later it's going to bitrot.

But you're the boss. If you want to keep it around, I'll rename the 
things and add some tests.

- Heikki

**Comment by vraghavan78** [comment]
Great, will plus 1 it after your updates.

**Comment by hlinnaka** [comment]
Ok, here's a new version. Renamed the things, and added tests.

**Comment by hlinnaka** [comment]
Ok, committed, thanks for the reviews!

**Comment by hlinnaka** [comment]
See https://github.com/greenplum-db/gpdb/pull/3594 for the GPDB changes to use the flag.

**Pull request merged**